### PR TITLE
util: Add userfaultfd memory monitor utility functions

### PR DESCRIPTION
**Pull request by wckzhang**
Existing glibc malloc hooks prove insufficient in detecting events on a
memory region. With the current state of the memory registration cache
component, if memory is freed through, for example, munmap, the cache
relying on these hooks will preserve stale entries. This will cause
inappropriate behavior in which sends will be performed through
unregistered buffers with stale mr keys. To resolve this invalidation
error, we can instead use userfault fd event handling which does support
munmap detection. This provides better coverage of memory events for the
memory registration cache component.

This commit implements common utility functions to handle userfault fd
creation, registration, unregistration, and cleanup.

Signed-off-by: William Zhang <wilzhang@amazon.com>

### CODE REVIEW

**Comment by wckzhang** [comment]
Note: I haven't been able to figure out a better way to detect a region's page size when registering/unregistering.

File: include/ofi_uffd.h [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+			"sysconf/pagesize: ret %ld, err %s\n",
+			ret, strerror(errno));
+		if (errno)
+			return -errno;
+		else
+			return -FI_EOTHER;
+	}
+
+	return ret;
+}
```
Comment (shefty) [inline include/ofi_uffd.h]: This function should be moved into a separate commit and different header file - maybe ofi_mem.h.  Please grep the code to see if this already exists somewhere else.

Comment (wckzhang) [reply include/ofi_uffd.h]: I don't think I found a function to do this. Will move to ofi_mem.h

File: include/ofi_uffd.h [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+#define MB(x)((size_t)(x) << 20)
+#define GB(x)((size_t)(x) << 30)
+
+/* See TODO in function ofi_uffd_register */
+static long page_sizes[] = {4096, MB(2), GB(1)};
```
Comment (jsquyres) [inline include/ofi_uffd.h]: This will create this symbol in each `.o` file.  Seems like it would be better in a `.c` file.

File: include/ofi_uffd.h [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	struct uffdio_register uffdio_register;
+	long page_size;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(page_sizes); i++) {
+		page_size = page_sizes[i];
+		set_uffdio_range(&uffdio_register.range, start, len, page_size);
+		uffdio_register.mode = UFFDIO_REGISTER_MODE_MISSING;
+
+		if (ioctl(uffd, UFFDIO_REGISTER, &uffdio_register) != -1)
```
Comment (jsquyres) [inline include/ofi_uffd.h]: Is the intent here that you repeatedly call `ioctl()` until with the different page sizes until it succeeds?

A comment about the intent of this loop would be appreciated.  😄

File: include/ofi_uffd.h [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+static inline int ofi_uffd_unregister(int uffd, uint64_t start, uint64_t len)
+{
+	struct uffdio_range uffdio_range;
+	long page_size;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(page_sizes); i++) {
+		page_size = page_sizes[i];
+		set_uffdio_range(&uffdio_range, start, len, page_size);
+		if (ioctl(uffd, UFFDIO_UNREGISTER, &uffdio_range) != -1)
```
Comment (jsquyres) [inline include/ofi_uffd.h]: I'm not overly familiar with the userfaultd stuff; forgive me if this is a naieve question.

Is it possible that you register region X, but then region X gets unmapped/removed from the process, but then before `ofi_uffd_unregister()` can get called to unregister region X, region X gets mapped again to some other physical address?

I.e., is there atomicity in the notifications of unmapping and the unregistration in this function?

File: configure.ac [hunk 5]
```diff
@@ -305,6 +306,42 @@ AC_CHECK_HEADER(valgrind/memcheck.h, [],
...
+have_userfaultfd=0
+AC_CHECK_HEADERS([linux/userfaultfd.h],
+    [AC_MSG_CHECKING([for userfaultfd interface])
+     AC_TRY_LINK([#include <sys/types.h>
+#include <linux/userfaultfd.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <fcntl.h>
+],
+        [return syscall(__NR_userfaultfd, O_CLOEXEC | O_NONBLOCK);],
```
Comment (jsquyres) [inline configure.ac]: FWIW, `AC_TRY_LINK` is deprecated -- `AC_LINK_IFELSE` is preferred these days.

That being said, do you need to try to link?  You're really just checking for all those header files and `__NR_userfaultfd`, right?  Because `syscall()` will definitely exists, as well `O_CLOEXEC` and `O_NONBLOCK`, right?  I.e., is there a reason not to use `AC_COMPILE_IFELSE`?  Or even just `AC_CHECK_HEADERS` and `AC_CHECK_DECL` (for `__NR_userfaultfd`)?

Perhaps something like (this is off the top of my head / not tested / YMMV):

```sh
have_userfault_fd=0
have_userfault_fd_msg=no
AC_MSG_CHECKING([for userfaultfd interface])
AC_CHECK_HEADER([linux/userfaultfd.h],
    [AC_CHECK_DECL([__NR_userfaultfd], [have_userfault_fd=1 have_userfault_fd_msg=yes])])
AC_MSG_RESULT([$have_userfault_fd_msg])
```

(same comments apply to your use of `AC_TRY_LINK` below)

File: configure.ac [hunk 6]
```diff
@@ -305,6 +306,42 @@ AC_CHECK_HEADER(valgrind/memcheck.h, [],
...
+            [AC_MSG_RESULT([yes])
+             have_userfaultfd=1],
+            [AC_MSG_RESULT([no])
+             have_userfaultfd=0])])
+AS_IF([test "$have_userfaultfd" -eq "1"]
```
Comment (jsquyres) [inline configure.ac]: `-eq` is a numeric comparison.  You should probably just have:

```
AS_IF([test $have_userfaultfd -eq 1],
```

(i.e., don't "case" the numbers into strings)

You're also missing a `,` at the end of this line.

File: include/ofi_uffd.h (hunk 4 above)
Comment (shefty) [reply include/ofi_uffd.h]: This is the general operation:
"When a page fault occurs for one of the regions registered to the userfaultfd object, the faulting thread is put to sleep and an event is generated that can be read via the userfaultfd file descriptor...  the fault-handling thread can trigger a wake-up for the sleeping thread."

So...you need a separate thread monitoring the fd.  Whenever a thread maps or unmaps memory, that thread will block in the kernel until the monitoring thread handles the event and wakes it back up.

File: include/ofi_uffd.h (hunk 3 above)
Comment (wckzhang) [reply include/ofi_uffd.h]: Yup, will add

File: include/ofi_uffd.h (hunk 4 above)
Comment (jsquyres) [reply include/ofi_uffd.h]: Probably also need to make 100% sure that the thread doesn't use any registered memory that may become unmapped (because then it will go to sleep).  This may be an obvious "duh", but I figured it was worth being explicit about... 😄

File: configure.ac (hunk 5 above)
Comment (bwbarrett) [reply configure.ac]: Yeah, ok, some of us are old and still use TRY_LINK, but good call out.

I think AC_CHECK_DECL is probably ok, I can't remember why I wanted to link instead of check the declaration.

File: configure.ac [hunk 7]
```diff
@@ -305,6 +306,42 @@ AC_CHECK_HEADER(valgrind/memcheck.h, [],
...
+    [AC_MSG_CHECKING([for userfaultfd unmap support])
+     AC_TRY_LINK([#include <sys/types.h>
```
Comment (bwbarrett) [inline configure.ac]: This one can be AC_COMPILE_IFELSE, but probably don't want a CHECK_DECL.  You'd need to check a bunch of different declarations, and then OR the results, and that doesn't seem useful.

File: configure.ac (hunk 6 above)
Comment (bwbarrett) [reply configure.ac]: huh, how did that work...

**Comment by jswaro** [comment]
Nice.

File: include/ofi_uffd.h [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+{
+	range->start = (start / page_size) * page_size;
+	range->len = ofi_div_ceil(len, page_size) * page_size;
+}
+
+/*
+ * TODO: ioctl register needs to be passed page aligned boundaries.
+ * Need to find a more sane way of detecting a region's page size.
+ */
+static inline int ofi_uffd_register(int uffd, uint64_t start, uint64_t len)
```
Comment (shefty) [inline include/ofi_uffd.h]: From the perspective of the rest of the code, I would anticipate start as void* and len as size_t.

**Comment by shefty** [comment]
I'd need to see how this will be integrated with the existing code to understand the intended use more.  I think that part of merging this will be removing the existing malloc/free hooks.  Those don't work in practice, so I don't believe we want to try maintaining them.  The current code writes events into a notification queue, that the cache processes when registering memory.  We need to see if that makes sense, or if we just want a callback model that updates the cache.  Basically, we can look at re-architecting pieces of what's there as needed.

I think getting this integrated is a fairly high priority.  If it helps, I can look at trying to add this support to the existing components.

File: include/ofi_uffd.h (hunk 8 above)
Comment (wckzhang) [reply include/ofi_uffd.h]: The range struct looks like this
```
struct uffdio_range {
	__u64 start;
	__u64 len;
};
```

I wanted to avoid casting it inside this file.

Comment (shefty) [reply include/ofi_uffd.h]: Why?  The cast is needed somewhere.  I would have expected this code to abstract away the userfaultfd details from the rest of the infrastructure, but this is exposing those details out.

File: prov/util/src/util_uffd.c [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+/* See TODO in function ofi_uffd_register */
+static long page_sizes[] = {4096, MB(2), GB(1)};
+
+static inline void set_uffdio_range(struct uffdio_range *range,
+				    void *start, size_t len,
+				    long page_size)
+{
+	range->start = ((uint64_t)start / page_size) * page_size;
+	range->len = ofi_div_ceil(len, page_size) * page_size;
+}
```
Comment (shefty) [inline prov/util/src/util_uffd.c]: I believe the above is incorrect.  If start is not page aligned, an extra page may be needed.  The referenced patch contains helper functions that I plan on using for this:

https://github.com/ofiwg/libfabric/pull/4956/commits/b15c579e1cda4ebbcec8ce6eb8734aa50816a0e2

It would be good to get a review on those changes, as they are critical to get the correct ranges.

**Comment by shefty** [comment]
I believe the functionality of this PR is now captured fully by PR #4956, though in a slightly different form.  Continuing with review there.

File: prov/util/src/util_uffd.c (hunk 9 above)
Comment (wckzhang) [reply prov/util/src/util_uffd.c]: If start is not page aligned, it will round down to the nearest page. I will review your patch

Comment (shefty) [reply prov/util/src/util_uffd.c]: My concern is that len is calculated incorrectly.
